kinto/core/storage/postgresql/__init__.py

```diff
@@ -701,43 +702,72 @@ def _format_conditions(self, filters, id_field, modified_field,
                     # Safely escape field name
                     field_holder = '{}_field_{}_{}'.format(prefix, i, j)
                     holders[field_holder] = subfield
-                    # Use ->> to convert the last level to text.
-                    column_name += "->>" if j == len(subfields) - 1 else "->"
+                    # Use ->> to convert the last level to text if
+                    # needed for LIKE query. (Other queries do JSONB comparison.)
+                    column_name += "->>" if j == len(subfields) - 1 and query_is_like else "->"
                     column_name += ":{}".format(field_holder)
 
-                # If field is missing, we default to ''.
-                sql_field = "coalesce({}, '')".format(column_name)
-                # Cast when comparing to number (eg. '4' < '12')
-                try:
-                    if value and (is_numeric(value) or all([is_numeric(v) for v in value])):
-                        sql_field = "({})::numeric".format(column_name)
-                except TypeError:  # not iterable
-                    pass
-
-            if filtr.operator not in (COMPARISON.IN, COMPARISON.EXCLUDE):
-                # For the IN operator, let psycopg escape the values list.
-                # Otherwise JSON-ify the native value (e.g. True -> 'true')
-                if not isinstance(value, str):
-                    value = json.dumps(value).strip('"')
-            else:
-                # If not all numeric, fallback to string comparison.
-                if not all([is_numeric(v) for v in value]):
-                    value = [str(v) for v in value]
+                # If the field is missing, column_name will produce
+                # NULL. NULL has strange properties with comparisons
+                # in SQL -- NULL = anything => NULL, NULL <> anything => NULL.
+                # We generally want missing fields to be treated as a
+                # special value that compares as different from
+                # everything, including JSON null. Do this on a
+                # per-operator basis.
+                if filtr.operator in (
+                        # NULLs aren't EQ to anything (definitionally).
+                        COMPARISON.EQ,
+                        # So they can't match anything in an INCLUDE.
+                        COMPARISON.IN,
+                        # Nor can they be LIKE anything.
+                        COMPARISON.LIKE,
+                ):
```

Comment:
**µnit**: in python multi-lines `if` are not recommended. Would it be better to have two intermediary variables like `null_comparable_ops` and `null_incomparable_ops` ?